### DISCUSSION TITLE
Using preprocessors on unstructured data

### DISCUSSION THREAD

**Discussion by schlunma** [ESMValGroup/ESMValCore#1643] [General]
In this discussion I want to collect information about how our preprocessors perform on unstructured data. This is motivated by the fact that many of them currently don't work on this kind of data.

Below I will collect a quick overview. Each preprocessor that has been tested is added with a box, the box is ticked if it works directly on the unstructured data (without prior regridding!). For preprocessors that do not work, a new discussion item can be opened to discuss progress.

Fell free to add new preprocessors if you discover that they don't work! Thanks!

- [x] `area_statistics` (fails with `ValueError: Invalid operation for 'latitude', with 3 bound(s)`)
- [x] `climate_statistics`
- [x] `extract_levels`
- [ ] `extract_region` (fails with `ValueError: expected 0 or 2 bound values per cell`)
- [x] `regrid` (when using appropriate schemes like `unstructured_nearest`)
- [x] `zonal_statistics` (fails with `ValueError: Invalid operation for 'latitude', with 3 bound(s)`)

**Comment by schlunma** [comment]
### `area_statistics` and `zonal_statistics`

#### Fails with

```
ValueError: Invalid operation for 'latitude', with 3 bound(s). Contiguous bounds are only defined for 1D coordinates with 2 bounds.
```

#### Debug log

[main_log_debug_area.txt](https://github.com/ESMValGroup/ESMValCore/files/8974443/main_log_debug.txt)
[main_log_debug_zonal.txt](https://github.com/ESMValGroup/ESMValCore/files/8974572/main_log_debug.txt)


#### Problem description

This is an issue of iris that is already reported here: https://github.com/SciTools/iris/issues/4672

#### Fix

This needs to be fixed in iris. I will open a PR as soon as I have the confirmation that the proposed solution in https://github.com/SciTools/iris/issues/4672 is appropriate.

**Comment by schlunma** [comment]
###  `extract_region`

#### Fails with

```
ValueError: expected 0 or 2 bound values per cell
```

#### Debug log

[main_log_debug.txt](https://github.com/ESMValGroup/ESMValCore/files/8974581/main_log_debug.txt)

#### Problem description

[`iris.cube.Cube.intersection`](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/cube.html#iris.cube.Cube.intersection) used [here](https://github.com/ESMValGroup/ESMValCore/blob/64ca458b6866dd699f09255d1e1ca325e8164349/esmvalcore/preprocessor/_area.py#L66-L70) cannot handle unstructured grids.

#### Fix

This could be solved by using [`iris.cube.Cube.extract`](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/cube.html#iris.cube.Cube.extract) with an appropriate constraint like

```py
iris.Constraint(latitude=lambda cell: start_latitude < cell.point < end_latitude) &
iris.Constraint(longitude=lambda cell: start_longitude < cell.point < end_longitude)
```
which would most likely also work for regular grids.

Is there any particular reason we use `Cube.intersection` here? Does anyone know @ESMValGroup/esmvaltool-coreteam ?

**Reply by bouweandela** [reply]
From the iris documentation, it looks like `intersection` has special support for the fact that latitude and longitude are spherical coordinates. I did not see any mention of this in the documentation of the `extract` method.

I also had some discussion with the iris developers around extracting regions in https://github.com/SciTools/iris/issues/4197. However, if we could just use `extract` that may not be needed at all?

**Reply by schlunma** [reply]
I haven't tested it for multi-dimensional coordinates, but for regular grids and unstructured grids (both have 1D coordinates) `extract` works fine. I think we should definitely consider replacing `intersection` with `extract`.

**Reply by bouweandela** [reply]
Did you also test more difficult cases? Such as start_longitude: -10, end_longitude: 10? Or even start_longitude: 10, end_longitude: -10?

**Reply by schlunma** [reply]
Ah, I didn't. In this case, `extract` either gives only the values between 0 and 10 (for `start_longitude: -10`, `end_longitude: 10`) or no cube at all (for `start_longitude: 10`, `end_longitude: -10`) as expected for a longitude that uses values between 0 and 360.

Our current preprocessor gives a cube with lons in [0, 10] and [350, 360] for the first case and an error (`ValueError: minimum greater than maximum`) for the second case.

Thus, `intersect` is a bit smarter than `extract` in terms of circular coordinates.

**Reply by bouweandela** [reply]
I think it would be best to open an issue in the iris repo about this and ask: why are there two ways to extract a region and why does it not work for all grids/regions?

**Reply by zklaus** [reply]
There are two things that set `intersection` apart from `extract`:
- `extract` works with arbitrary `Constraint`s, `intersection` specifically with coordinate ranges
- `intersection` rolls circular coordinates, `extract` simply extracts

I think it's reasonably clear :shrug:

**Reply by schlunma** [reply]
@zklaus That kind of makes sense, it's only weird that it also works for `latitude` and not only for `longitude`.

In this case, it would be great if you could [upvote](https://github.com/SciTools/iris/issues/4631) the issue referenced by Bouwe (https://github.com/SciTools/iris/issues/4197) with a :+1:.

**Reply by zklaus** [reply]
Done.
